MapboxCoreNavigation/RouteController.swift

```diff
@@ -194,28 +191,36 @@ open class RouteController: NSObject {
         checkForFasterRoute(from: location, routeProgress: routeProgress)
     }
     
-    func updateSpokenInstructionProgress(status: MBNavigationStatus, willReRoute: Bool) {
+    func updateIndexes(status: MBNavigationStatus) {
+        routeProgress.legIndex = Int(status.legIndex)
+        routeProgress.currentLegProgress.stepIndex = Int(status.stepIndex)
+        routeProgress.currentLegProgress.currentStepProgress.intersectionIndex = Int(status.intersectionIndex)
         
-        if let voiceInstructionIndex = status.voiceInstruction?.index {
-            routeProgress.currentLegProgress.currentStepProgress.spokenInstructionIndex = Int(voiceInstructionIndex)
-            
-            // Don't annouce spoken instruction if we are going to reroute
-            if !willReRoute,
-                let spokenInstruction = routeProgress.currentLegProgress.currentStepProgress.currentSpokenInstruction {
-                announcePassage(of: spokenInstruction, routeProgress: routeProgress)
-            }
+        if let spokenInstructionIndex = status.voiceInstruction?.index {
+            routeProgress.currentLegProgress.currentStepProgress.spokenInstructionIndex = Int(spokenInstructionIndex)
+        }
+        
+        if let visualInstructionIndex = status.bannerInstruction?.index {
+            routeProgress.currentLegProgress.currentStepProgress.visualInstructionIndex = Int(visualInstructionIndex)
+        }
+    }
+    
+    func updateSpokenInstructionProgress(status: MBNavigationStatus, willReRoute: Bool) {
+        let didUpdate = status.voiceInstruction?.index != nil
+
+        // Announce voice instruction if it was updated and we are not going to reroute
+        if didUpdate && !willReRoute,
+            let spokenInstruction = routeProgress.currentLegProgress.currentStepProgress.currentSpokenInstruction {
+            announcePassage(of: spokenInstruction, routeProgress: routeProgress)
         }
     }
     
     func updateVisualInstructionProgress(status: MBNavigationStatus) {
+        let didUpdate = status.bannerInstruction != nil
         
-        let willChangeVisualIndex = status.bannerInstruction != nil
-        
-        if willChangeVisualIndex || isFirstLocation {
-            let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
-            currentStepProgress.visualInstructionIndex = Int(status.bannerInstruction?.index ?? 0)
-            
-            if let instruction = currentStepProgress.currentVisualInstruction {
+        // Announce visual instruction if it was updated or it is the first location being reported
+        if didUpdate || isFirstLocation {
```

Comment:
If `didUpdate` is false, then is it safe to announce anything even if `isFirstLocation` is true?

Reply:
yep

Reply:
if `isFirstLocation` is true, that would mean `stepIndex`, `visualInstructionIndex`, and `spokenInstructionIndex` is 0. I can't think if a case where that wouldn't be safe, besides routeless navigation which currently is not supported.

Reply:
Ah, between the whitespace changes and variable renaming, I had mistakenly thought the logic had changed too. 👍

Incidentally, how well do we handle the pathological case of a route that ends at exactly or roughly the same spot as it begins, or a route with two coincident destinations? If I remember correctly, one of the features of OSRM is that it can give you a no-op route, unlike some other routing engines that make you go in a circle. Perhaps that edge case raises other issues that are out of scope for this PR, but that’s the only case that seems relevant to me.

Reply:
> case of a route that ends at exactly or roughly the same spot as it begins, or a route with two coincident destinations?

Valid concerns but slightly out of scope for this PR.

I found one more place where indexes could be out of sync.

https://github.com/mapbox/mapbox-navigation-ios/blob/b309653a8acf26dcececcde4848e3060fa4ffac1/MapboxCoreNavigation/RouteController.swift#L253-L257

Fixed in https://github.com/mapbox/mapbox-navigation-ios/pull/2138/commits/bfe5ba7db6eec2e188ca33b7de386f8ec640f7ba